BuoyantIO/linkerd-buoyant
# Add API to transfer Linkerd certs data

This change augments the agent's functionality to send Linkerd TLS cert data
to the bcloud api. This is done through a new API endpoint called `LinkerdInfo`.
We envision that this API will serve to transfer data other than certificates in the 
future. For now we can transfer three type of certs objects - control plane certs, 
proxy certs and webhook certs. This PR adds implementation to send only control
plane certificates.

To test: 
1. Bring up your docker compose env.
2. Create a kind cluster and install Linkerd on it.
3. Build the agent
`docker build -f agent/Dockerfile -t ghcr.io/buoyantio/linkerd-buoyant:certs .`
4. Load the image onto the kind cluster
 `kind load docker-image ghcr.io/buoyantio/linkerd-buoyant:certs`
5. Add a cluster
6. Get your agent manifest, but do not apply it as you need to make some edits
7. Edit the agent image to point to `linkerd-buoyant:certs` and the pull policy to `IfNotPresent`
8. Add a ports skip annotation to the bcloud agent deployment spec.
That needs to go in `deployment.spec.template.metadata.annotations`
  ```
  config.linkerd.io/skip-outbound-ports: 4191
  ```
9. Now you can apply the agent manifest
10. Tail the log of the agent: 

```
 kubectl logs -n buoyant-cloud <agent-pod> buoyant-cloud-agent  -f | grep handleLinkerdInfo
```

Note that this will not be very interesting because the default LD installation makes the roots
and the issuer the same: https://github.com/linkerd/linkerd2/blob/main/cli/cmd/options.go#L673. 
If you want to make the experience more exciting, you can supply your own certs or even configure
cert manager to rotate them periodically

Signed-off-by: Zahari Dichev <zaharidichev@gmail.com>
